client/src/components/EmojiPicker.js

```diff
@@ -0,0 +1,169 @@
+            this.setState({ keysDown: keysDown });
+        }
+    }
+    onKeyDown(event) {
+        if (event.which === 8) {
+            // Backspace
+            this.selectEmoji(-1);
+            if(this.state.highlightedIndex !== 0) {
+                event.preventDefault();
+                this.setState({ highlightedIndex: 0});
+            }
+        } else if (event.which === 16) {
+            // Shift
+            const keysDown = { 'shift' : true };
+            this.setState({ keysDown: keysDown });
+        } else if(event.which === 9) {
+            // Tab
+            event.preventDefault();
+            if (this.state.keysDown['shift']) {
+                // Shift + Tab
+                this.cycle(-1);
+            } else {
+                this.cycle(1);
+            }
+        } else if (event.which === 39) {
+            // Right arrow
+            event.preventDefault();
+            this.cycle(1);
+        } else if (event.which === 37) {
+            // Left arrow
+            event.preventDefault();
+            this.cycle(-1);
+        } else if (event.which === 38) {
+            // Up arrow
+            event.preventDefault();
+            const step = -this.props.elemsPerRow;
+            this.cycleRow(step);
+        } else if (event.which === 40) {
+            // Down arrow
+            event.preventDefault();
+            const step = this.props.elemsPerRow;
+            this.cycleRow(step);
+        } else if (event.which === 13 || event.which === 32) {
+            // Return or Space
+            event.preventDefault();
+            this.selectEmoji(this.state.highlightedIndex);
+            this.props.onClose();
+        } else if (event.which === 27) {
+            //ESC
+            this.selectEmoji(-1);
+            this.props.onClose();
+          }
+      }
+
+      onClick(index) {
+          this.setState({ highlightedIndex: index});
+          this.selectEmoji(index);
+          this.props.onClose();
+      }
+
+      onMouseEnter(index) {
+          this.setState({ highlightedIndex: index});
+          this.selectEmoji(index);
+      }
+
+      cycleRow(step) {
+          const newIndex = this.state.highlightedIndex + step;
+          const elemsPerRow = this.props.elemsPerRow;
+          const offset = modulo(newIndex, elemsPerRow);
+          const count = this.state.emojis.length;
+          const numRows = Math.ceil(count/elemsPerRow);
+          const maxRow = numRows*elemsPerRow - (elemsPerRow - offset) < count ? numRows : numRows - 1;
+          const row = modulo(Math.floor(newIndex / elemsPerRow), maxRow);
+          const highlightedIndex = row*elemsPerRow + offset;
+          this.setState({ highlightedIndex: highlightedIndex });
+          this.selectEmoji(highlightedIndex);
+      }
+
+      cycle(step) {
+          const newIndex = this.state.highlightedIndex + step;
+          const count = this.state.emojis.length;
+          const highlightedIndex = modulo(newIndex, count);
```

Comment:
Since the variables here are used only in one place, I think you can combine the three rows here to:
```javascript
cycle(step) {
  const highlightedIndex = modulo(this.state.highlightedIndex + step, this.state.emojis.length);
  this.setState({ highlightedIndex: highlightedIndex });
  this.selectEmoji(highlightedIndex);
}

```